streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -38,346 +40,626 @@
+
+        private final TopicPartition changelogPartition;
+
+        private final ProcessorStateManager stateManager;
+
+        private ChangelogState changelogState;
+
+        private long totalRestored;
+
+        // only for active restoring tasks (for standby it is null)
+        private Long restoreEndOffset;
+
+        // only for standby tasks that use source topics as changelogs (for active it is null);
+        // if it is not on source topics it is also null
+        private Long restoreLimitOffset;
+
+        // only for standby tasks to buffer records polled but exceed limit at the moment;
+        // for active it should be null
+        private List<ConsumerRecord<byte[], byte[]>> bufferedRecords;
+
+        // NOTE we do not book keep the current offset since we leverage state manager as its source of truth
+
+        private ChangelogMetadata(final TopicPartition changelogPartition, final ProcessorStateManager stateManager) {
+            final boolean shouldUpdateLimitOffset = stateManager.changelogAsSource(changelogPartition);
+
+            this.stateManager = stateManager;
+            this.changelogPartition = changelogPartition;
+            this.changelogState = ChangelogState.REGISTERED;
+            this.restoreEndOffset = null;
+            this.totalRestored = 0L;
+
+            if (shouldUpdateLimitOffset) {
+                this.bufferedRecords = new ArrayList<>();
+                this.restoreLimitOffset = 0L;
+            } else {
+                this.bufferedRecords = null;
+                this.restoreLimitOffset = null;
+            }
+        }
+
+        private void clear() {
+            if (this.bufferedRecords != null) {
+                this.bufferedRecords.clear();
+                this.bufferedRecords = null;
+            }
+        }
+
+        @Override
+        public String toString() {
+            return changelogState + " (endOffset " + restoreEndOffset + ", limitOffset " + restoreLimitOffset + ")";
+        }
+    }
+
+    private ChangelogReaderState state;
+
     private final Logger log;
+    private final Duration pollTime;
+
+    // 1) we keep adding partitions to restore consumer whenever new tasks are registered with the state manager;
+    // 2) we do not unassign partitions when we switch between standbys and actives, we just pause / resume them;
+    // 3) we only remove an assigned partition when the corresponding task is being removed from the thread.
     private final Consumer<byte[], byte[]> restoreConsumer;
-    private final StateRestoreListener userStateRestoreListener;
-    private final Map<TopicPartition, Long> restoreToOffsets = new HashMap<>();
+    private final StateRestoreListener stateRestoreListener;
+
+    // source of the truth of the current registered changelogs;
+    // NOTE a changelog would only be removed when its corresponding task
+    // is being removed from the task; otherwise it would stay in this map even after completed
```

Comment:
"its corresponding task is being removed from the **task**" --> "from the **thread**"?

Reply:
ack